cpp/tests/utilities/cxxopts.hpp

```diff
@@ -0,0 +1,1497 @@
+  std::shared_ptr<const Value> value_;
+  std::string arg_help_;
+};
+
+class Options {
+  typedef std::unordered_map<std::string, std::shared_ptr<OptionDetails>> OptionMap;
+
+ public:
+  Options(std::string program, std::string help_string = "")
+    : m_program(std::move(program)),
+      m_help_string(toLocalString(std::move(help_string))),
+      m_custom_help("[OPTION...]"),
+      m_positional_help("positional parameters"),
+      m_show_positional(false),
+      m_allow_unrecognised(false),
+      m_options(std::make_shared<OptionMap>()),
+      m_next_positional(m_positional.end())
+  {
+  }
+
+  Options& positional_help(std::string help_text)
+  {
+    m_positional_help = std::move(help_text);
+    return *this;
+  }
+
+  Options& custom_help(std::string help_text)
+  {
+    m_custom_help = std::move(help_text);
+    return *this;
+  }
+
+  Options& show_positional_help()
+  {
+    m_show_positional = true;
+    return *this;
+  }
+
+  Options& allow_unrecognised_options()
+  {
+    m_allow_unrecognised = true;
+    return *this;
+  }
+
+  ParseResult parse(int& argc, char**& argv);
+
+  OptionAdder add_options(std::string group = "");
+
+  void add_options(const std::string& group, std::initializer_list<Option> options);
+
+  void add_option(const std::string& group, const Option& option);
+
+  void add_option(const std::string& group,
+                  const std::string& s,
+                  const std::string& l,
+                  std::string desc,
+                  std::shared_ptr<const Value> value,
+                  std::string arg_help);
+
+  // parse positional arguments into the given option
+  void parse_positional(std::string option);
+
+  void parse_positional(std::vector<std::string> options);
+
+  void parse_positional(std::initializer_list<std::string> options);
+
+  template <typename Iterator>
+  void parse_positional(Iterator begin, Iterator end)
+  {
+    parse_positional(std::vector<std::string>{begin, end});
+  }
+
+  std::string help(const std::vector<std::string>& groups = {}) const;
+
+  const std::vector<std::string> groups() const;
+
+  const HelpGroupDetails& group_help(const std::string& group) const;
+
+ private:
+  void add_one_option(const std::string& option, std::shared_ptr<OptionDetails> details);
+
+  String help_one_group(const std::string& group) const;
+
+  void generate_group_help(String& result, const std::vector<std::string>& groups) const;
+
+  void generate_all_groups_help(String& result) const;
+
+  std::string m_program;
+  String m_help_string;
+  std::string m_custom_help;
+  std::string m_positional_help;
+  bool m_show_positional;
+  bool m_allow_unrecognised;
+
+  std::shared_ptr<OptionMap> m_options;
+  std::vector<std::string> m_positional;
+  std::vector<std::string>::iterator m_next_positional;
+  std::unordered_set<std::string> m_positional_set;
+
+  // mapping from groups to help options
+  std::map<std::string, HelpGroupDetails> m_help;
+};
+
+class OptionAdder {
```

Comment:
Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`.